pilot/pkg/networking/core/v1alpha3/envoyfilter/listener_patch.go

```diff
@@ -531,28 +567,107 @@ func patchHTTPFilters(patchContext networking.EnvoyFilter_PatchContext,
 	}
 }
 
+// orderDependencies recursively orders the patches based on insert before/after positions.
+func orderDependencies(patch *model.EnvoyFilterConfigPatchWrapper, original []*model.EnvoyFilterConfigPatchWrapper,
+	ordered *[]*model.EnvoyFilterConfigPatchWrapper) {
+	for _, lp := range original {
+		if lp == patch {
+			continue
+		}
+		if (patch.Operation == networking.EnvoyFilter_Patch_INSERT_BEFORE && hasHTTPFilterMatch(patch)) ||
+			(patch.Operation == networking.EnvoyFilter_Patch_INSERT_AFTER && hasHTTPFilterMatch(patch)) {
+			filter := lp.Value.(*http_conn.HttpFilter)
+			match := patch.Match.GetListener().FilterChain.Filter.SubFilter
+			if match != nil && nameMatches(match.Name, filter.Name) {
+				orderDependencies(lp, original, ordered)
+			}
+		}
+	}
+	if !contains(*ordered, patch) {
+		*ordered = append(*ordered, patch)
+	}
+}
+
+func contains(patches []*model.EnvoyFilterConfigPatchWrapper, patch *model.EnvoyFilterConfigPatchWrapper) bool {
+	for _, p := range patches {
+		if p == patch {
+			return true
+		}
+	}
+	return false
+}
+
+// hasCyclicDependencies checks if the patches have cyclic dependencies.
+// Inspired by https://github.com/dnaeon/go-dependency-graph-algorithm/blob/master/dependency-graph.go#L56.
+func hasCyclicDependencies(original []*model.EnvoyFilterConfigPatchWrapper) bool {
+	dependencies := make(map[string]sets.Set)
+	// First collect all dependencies for every patch.
+	for _, lp := range original {
+		lfilter := lp.Value.(*http_conn.HttpFilter)
+		deps := sets.NewSet()
+		for _, rp := range original {
+			if lp == rp {
+				continue
+			}
+			rmatch := rp.Match.GetListener().FilterChain.Filter.SubFilter
+			if rmatch != nil && nameMatches(rmatch.Name, lfilter.Name) {
+				deps.Insert(rp.Value.(*http_conn.HttpFilter).Name)
+			}
+		}
+		dependencies[lfilter.Name] = deps
+	}
+
+	// Iteratively find and remove patches which have no dependencies. If at some point
+	// there are still patches in the dependencies and we cannot find patches without
+	// dependencies, that means we have a circular dependency.
+	for len(dependencies) != 0 {
+		// Get all filters from the dependencies which have no dependencies
+		independent := sets.NewSet()
+		for name, deps := range dependencies {
```

Comment:
I think the name is confusing here. Either I am not reading it correctly or the name is incorrect.

“””
for filtername, dependents := range dependencies {
 “””
If len(dependents) == 0 it does not make the filter independent.
The cycle detection may still be correct.

Reply:
looks like dependencies is causing confusing. Let me rename that.